spynnaker/pyNN/models/neural_projections/connectors/from_list_connector.py

```diff
@@ -38,52 +37,90 @@ def __init__(self, conn_list, safe=True, verbose=False):
                 " at least a list of tuples, each of which should contain:"
                 " (pre_idx, post_idx)")
         self._conn_list = conn_list
-
-        # supports setting these at different times
-        self._weights = None
-        self._delays = None
         self._converted_weights_and_delays = False
 
+    @staticmethod
+    def _split_conn_list(conn_list, column_names):
+        """ Separate the connection list into the blocks needed.
+        :param conn_list: the original connection list
+        :param column_names: the column names if exist
+        :return: source dest list, weights list, delays list, extra list
+        """
+
+        # weights and delay index
+        weight_index = None
+        delay_index = None
+
+        # conn lists
+        weights = None
+        delays = None
+
+        # locate weights and delay index in the listings
+        if "weight" in column_names:
+            weight_index = column_names.index("weight")
+        if "delay" in column_names:
+            delay_index = column_names.index("delay")
+        element_index = list(range(2, len(column_names)))
+
+        # figure out where other stuff is
+        conn_list = numpy.array(conn_list)
+        source_destination_conn_list = conn_list[:, [0, 1]]
+
+        if weight_index is not None:
+            element_index.remove(weight_index)
+            weights = conn_list[:, weight_index]
+        if delay_index is not None:
+            element_index.remove(delay_index)
+            delays = conn_list[:, delay_index]
+
+        # build other data element conn list (with source and destination)
+        other_conn_list = None
+        other_element_column_names = list()
+        for element in element_index:
+            other_element_column_names.append(column_names[element])
+        if element_index:
+            other_conn_list = conn_list[:, element_index]
+            other_conn_list.dtype.names = other_element_column_names
+
+        # hand over split data
+        return source_destination_conn_list, weights, delays, other_conn_list
+
     @overrides(AbstractConnector.set_weights_and_delays)
     def set_weights_and_delays(self, weights, delays):
```

Comment:
I think there's something in this that I still need to work out - namely that a user could supply only weights, or only delays (assuming that's allowed by PyNN, which does seem to be the case).  I actually worked on something similar to this in a later PR, so this could be dealt with in that PR (see https://github.com/SpiNNakerManchester/sPyNNaker8/pull/188 and https://github.com/SpiNNakerManchester/sPyNNaker/pull/586) rather than in this one?

Reply:
The above fix should work in general since you can check for weights and delays separately.  Basically, if you store the weights and/or delays from the FromListConnector constructor, you can then check if they have been provided when any functions are called, and use the ones provided in the list if you have them.